File: backend/routers/production.py
"""生产执行链 API"""
from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from database import get_db
from production_service import (
    barcode_to_dict,
    confirm_material_doc,
    create_fg_from_qa,
    create_material_doc,
    create_mo,
    create_mo_from_plan,
    create_qa_from_mo,
    fg_to_dict,
    get_fg,
    get_material_doc,
    get_mo,
    get_qa,
    judge_qa,
    list_barcodes,
    list_fg,
    list_material_docs,
    list_mos,
    list_qas,
    material_doc_to_dict,
    mo_to_dict,
    pmc_mo_board,
    post_fg,
    qa_to_dict,
    register_barcode,
    set_mo_status,
)
from system_auth import AuthPrincipal, require_system_auth

router = APIRouter(
    prefix="/api/production",
    tags=["production"],
    dependencies=[Depends(require_system_auth)],
)

PROD_ROLES = frozenset({"admin", "production", "pmc", "planner", "quality", "warehouse"})


def require_prod(principal: AuthPrincipal = Depends(require_system_auth)) -> AuthPrincipal:
    name = (principal.username or "").strip().lower()
    if principal.role in PROD_ROLES or name in {"wgq", "dx001"}:
        return principal
    raise HTTPException(status_code=403, detail="无生产模块权限")


class MoIn(BaseModel):
    material_code: str = ""
    material_name: str = ""
    qty: float = 0
    unit: str = "PCS"
    due_date: str = ""
    bom_model_id: Optional[int] = None
    source_so_id: Optional[int] = None
    source_so_no: str = ""
    stock_owner: str = "internal"
    remark: str = ""


class StatusIn(BaseModel):
    status: str


class FromPlanIn(BaseModel):
    line_id: Optional[int] = None


class BarcodeIn(BaseModel):
    barcode: str
    remark: str = ""


class MaterialLineIn(BaseModel):
    material_code: str = ""
    material_name: str = ""
    qty: float = 0
    unit: str = "PCS"
    remark: str = ""


class MaterialDocIn(BaseModel):
    mo_id: int
    kind: str = "issue"
    from_bom: bool = False
    remark: str = ""
    lines: list[MaterialLineIn] = Field(default_factory=list)


class JudgeQaIn(BaseModel):
    pass_qty: float = 0
    fail_qty: float = 0


class FgFromQaIn(BaseModel):
    direct_outbound: bool = False


def _dump(m: BaseModel) -> dict[str, Any]:
    return m.model_dump() if hasattr(m, "model_dump") else m.dict()  # type: ignore[attr-defined]


@router.get("/orders")
def api_list_mo(
    q: str = Query(""),
    status: str = Query(""),
    db: Session = Depends(get_db),
    _: AuthPrincipal = Depends(require_prod),
):
    return {"items": [mo_to_dict(r) for r in list_mos(db, q=q, status=status)]}


@router.post("/orders")
def api_create_mo(body: MoIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)):
    try:
        row = create_mo(db, _dump(body), user=principal.username)
        db.commit()
        return mo_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/orders/{mo_id}")
def api_get_mo(mo_id: int, db: Session = Depends(get_db), _: AuthPrincipal = Depends(require_prod)):
    try:
        return mo_to_dict(get_mo(db, mo_id))
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e)) from e


@router.post("/orders/{mo_id}/status")
def api_mo_status(
    mo_id: int, body: StatusIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        row = set_mo_status(db, mo_id, body.status, user=principal.username)
        db.commit()
        return mo_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.post("/orders/from-plan/{plan_id}")
def api_from_plan(
    plan_id: int, body: FromPlanIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        rows = create_mo_from_plan(db, plan_id, line_id=body.line_id, user=principal.username)
        db.commit()
        return {"items": [mo_to_dict(r) for r in rows]}
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.post("/orders/{mo_id}/barcodes")
def api_barcode(
    mo_id: int, body: BarcodeIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        row = register_barcode(db, mo_id, body.barcode, user=principal.username, remark=body.remark)
        db.commit()
        return barcode_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/orders/{mo_id}/barcodes")
def api_list_bc(mo_id: int, db: Session = Depends(get_db), _: AuthPrincipal = Depends(require_prod)):
    return {"items": [barcode_to_dict(r) for r in list_barcodes(db, mo_id)]}


@router.post("/materials")
def api_create_mat(
    body: MaterialDocIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        row = create_material_doc(
            db,
            mo_id=body.mo_id,
            kind=body.kind,
            lines=[_dump(x) for x in body.lines] if body.lines else None,
            from_bom=body.from_bom,
            user=principal.username,
            remark=body.remark,
        )
        db.commit()
        row, lines = get_material_doc(db, row.id)
        return material_doc_to_dict(row, lines)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/materials")
def api_list_mat(
    kind: str = Query(""),
    mo_id: Optional[int] = None,
    db: Session = Depends(get_db),
    _: AuthPrincipal = Depends(require_prod),
):
    items = []
    for r in list_material_docs(db, kind=kind, mo_id=mo_id):
        _, lines = get_material_doc(db, r.id)
        items.append(material_doc_to_dict(r, lines))
    return {"items": items}


@router.post("/materials/{doc_id}/confirm")
def api_confirm_mat(
    doc_id: int, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        confirm_material_doc(db, doc_id, user=principal.username)
        db.commit()
        row, lines = get_material_doc(db, doc_id)
        return material_doc_to_dict(row, lines)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.post("/qa/from-mo/{mo_id}")
def api_create_qa(mo_id: int, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)):
    try:
        row = create_qa_from_mo(db, mo_id, user=principal.username)
        db.commit()
        return qa_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/qa")
def api_list_qa(db: Session = Depends(get_db), _: AuthPrincipal = Depends(require_prod)):
    return {"items": [qa_to_dict(r) for r in list_qas(db)]}


@router.post("/qa/{qa_id}/judge")
def api_judge_qa(
    qa_id: int, body: JudgeQaIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        row = judge_qa(db, qa_id, pass_qty=body.pass_qty, fail_qty=body.fail_qty, user=principal.username)
        db.commit()
        return qa_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.post("/fg-receipts/from-qa/{qa_id}")
def api_fg_from_qa(
    qa_id: int, body: FgFromQaIn, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)
):
    try:
        row = create_fg_from_qa(db, qa_id, user=principal.username, direct_outbound=body.direct_outbound)
        db.commit()
        return fg_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/fg-receipts")
def api_list_fg(db: Session = Depends(get_db), _: AuthPrincipal = Depends(require_prod)):
    return {"items": [fg_to_dict(r) for r in list_fg(db)]}


@router.post("/fg-receipts/{rid}/post")
def api_post_fg(rid: int, db: Session = Depends(get_db), principal: AuthPrincipal = Depends(require_prod)):
    try:
        row = post_fg(db, rid, user=principal.username)
        db.commit()
        return fg_to_dict(row)
    except ValueError as e:
        db.rollback()
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/pmc-board")
def api_pmc(db: Session = Depends(get_db), _: AuthPrincipal = Depends(require_prod)):
    return {"items": pmc_mo_board(db)}

File: frontend/src/api/production.ts
import { apiFetch } from './http'

export type Mo = {
  id: number
  mo_no: string
  material_code: string
  material_name: string
  qty: number
  status: string
  bom_model_id?: number | null
  source_plan_no?: string
  source_so_no?: string
  issued_sets?: number
  qa_pass_qty?: number
  fg_qty?: number
}

export type MaterialDoc = {
  id: number
  doc_no: string
  kind: string
  mo_no: string
  status: string
  lines?: Array<{ material_code: string; material_name: string; qty: number }>
}

export type QaDoc = {
  id: number
  qa_no: string
  mo_no: string
  status: string
  result: string
  qty: number
  pass_qty: number
  fail_qty: number
}

export type FgDoc = {
  id: number
  receipt_no: string
  mo_no: string
  material_code: string
  qty: number
  status: string
  direct_outbound?: boolean
}

export function fetchMos(params: { q?: string; status?: string } = {}) {
  const qs = new URLSearchParams()
  if (params.q) qs.set('q', params.q)
  if (params.status) qs.set('status', params.status)
  const s = qs.toString()
  return apiFetch<{ items: Mo[] }>(`/production/orders${s ? `?${s}` : ''}`)
}

export function createMo(body: Partial<Mo> & { material_code: string; qty: number; bom_model_id?: number }) {
  return apiFetch<Mo>('/production/orders', { method: 'POST', body: JSON.stringify(body) })
}

export function setMoStatus(id: number, status: string) {
  return apiFetch<Mo>(`/production/orders/${id}/status`, { method: 'POST', body: JSON.stringify({ status }) })
}

export function createMoFromPlan(planId: number) {
  return apiFetch<{ items: Mo[] }>(`/production/orders/from-plan/${planId}`, {
    method: 'POST',
    body: JSON.stringify({}),
  })
}

export function registerMoBarcode(moId: number, barcode: string) {
  return apiFetch(`/production/orders/${moId}/barcodes`, {
    method: 'POST',
    body: JSON.stringify({ barcode }),
  })
}

export function createMaterialDoc(body: {
  mo_id: number
  kind: string
  from_bom?: boolean
  lines?: Array<{ material_code: string; material_name: string; qty: number }>
}) {
  return apiFetch<MaterialDoc>('/production/materials', { method: 'POST', body: JSON.stringify(body) })
}

export function fetchMaterialDocs(params: { kind?: string; mo_id?: number } = {}) {
  const qs = new URLSearchParams()
  if (params.kind) qs.set('kind', params.kind)
  if (params.mo_id) qs.set('mo_id', String(params.mo_id))
  const s = qs.toString()
  return apiFetch<{ items: MaterialDoc[] }>(`/production/materials${s ? `?${s}` : ''}`)
}

export function confirmMaterialDoc(id: number) {
  return apiFetch<MaterialDoc>(`/production/materials/${id}/confirm`, { method: 'POST' })
}

export function createQaFromMo(moId: number) {
  return apiFetch<QaDoc>(`/production/qa/from-mo/${moId}`, { method: 'POST' })
}

export function fetchQas() {
  return apiFetch<{ items: QaDoc[] }>('/production/qa')
}

export function judgeQa(id: number, pass_qty: number, fail_qty: number) {
  return apiFetch<QaDoc>(`/production/qa/${id}/judge`, {
    method: 'POST',
    body: JSON.stringify({ pass_qty, fail_qty }),
  })
}

export function createFgFromQa(qaId: number, direct_outbound = false) {
  return apiFetch<FgDoc>(`/production/fg-receipts/from-qa/${qaId}`, {
    method: 'POST',
    body: JSON.stringify({ direct_outbound }),
  })
}

export function fetchFgReceipts() {
  return apiFetch<{ items: FgDoc[] }>('/production/fg-receipts')
}

export function postFg(id: number) {
  return apiFetch<FgDoc>(`/production/fg-receipts/${id}/post`, { method: 'POST' })
}

export function fetchPmcBoard() {
  return apiFetch<{ items: Array<Record<string, unknown>> }>('/production/pmc-board')
}
